File: research_and_backtests/amd_backtest.zig
const std = @import("std");

const Candle = struct {
    time: i64,
    open: f64,
    high: f64,
    low: f64,
    close: f64,
};

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    const file = try std.fs.cwd().openFile("xauusd_clean.csv", .{});
    defer file.close();

    var reader = file.reader();
    var buf: [1024]u8 = undefined;
    
    var candles = std.ArrayList(Candle).init(allocator);
    defer candles.deinit();

    while (try reader.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        if (line.len == 0) continue;
        var it = std.mem.split(u8, line, ",");
        
        const t_str = it.next() orelse continue;
        const o_str = it.next() orelse continue;
        const h_str = it.next() orelse continue;
        const l_str = it.next() orelse continue;
        const c_str = it.next() orelse continue;

        const time = try std.fmt.parseInt(i64, t_str, 10);
        const o = try std.fmt.parseFloat(f64, o_str);
        const h = try std.fmt.parseFloat(f64, h_str);
        const l = try std.fmt.parseFloat(f64, l_str);
        const c = try std.fmt.parseFloat(f64, c_str);

        try candles.append(.{ .time = time, .open = o, .high = h, .low = l, .close = c });
    }

    std.debug.print("=========================================================\n", .{});
    std.debug.print("[*] ZIG ENGINE: Loaded {} M5 Candles (60 Days)\n", .{candles.items.len});

    // AMD Backtest Logic
    var balance: f64 = 10750.0; // ~$10,750 USD = 9 Lakh INR
    var max_balance: f64 = balance;
    var total_trades: i32 = 0;
    var wins: i32 = 0;
    
    var in_trade: bool = false;
    var trade_dir: i32 = 0; // 1 = buy, -1 = sell
    var entry_price: f64 = 0;
    var sl: f64 = 0;
    var tp: f64 = 0;
    var trade_volume: f64 = 0;

    var asian_high: f64 = 0.0;
    var asian_low: f64 = 99999.0;
    var has_traded_today: bool = false;
    var current_day: i64 = 0;

    for (candles.items) |c| {
        // Unix time approximation (seconds)
        const day = @divFloor(c.time, 86400);
        const hour = @mod(@divFloor(c.time, 3600), 24);

        if (day != current_day) {
            current_day = day;
            asian_high = 0.0;
            asian_low = 99999.0;
            has_traded_today = false;
            
            // Close trades overnight to avoid massive gap risk
            if (in_trade) {
                if (trade_dir == 1) {
                    balance += (c.open - entry_price) * trade_volume;
                } else {
                    balance += (entry_price - c.open) * trade_volume;
                }
                in_trade = false;
            }
        }

        // Asian Session (00:00 to 06:00 UTC) Accumulation
        if (hour >= 0 and hour < 6) {
            if (c.high > asian_high) asian_high = c.high;
            if (c.low < asian_low) asian_low = c.low;
        }

        // London/NY Session Manipulation & Distribution
        if (hour >= 6 and hour < 18 and !in_trade and !has_traded_today) {
            const range_size = asian_high - asian_low;
            
            // Only trade if Asian range is relatively tight (accumulation occurred)
            if (range_size > 2.0 and range_size < 15.0) {
                // Liquidity Sweep (Fakeout)
                if (c.high > asian_high and c.close < asian_high) {
                    // Sell Signal: Buy stops swept, reversing down
                    in_trade = true;
                    trade_dir = -1;
                    entry_price = c.close;
                    sl = c.high + 0.5; // Micro stop above the sweep wick
                    const risk_distance = sl - entry_price;
                    tp = entry_price - (risk_distance * 4.0); // 1:4 R/R Expansion
                    
                    // Kamikaze-Adjacent Risk: 2.5% per trade (safely under 3% limit)
                    const risk_amount = balance * 0.025;
                    trade_volume = risk_amount / risk_distance; 
                    has_traded_today = true;
                } else if (c.low < asian_low and c.close > asian_low) {
                    // Buy Signal: Sell stops swept, reversing up
                    in_trade = true;
                    trade_dir = 1;
                    entry_price = c.close;
                    sl = c.low - 0.5; // Micro stop below the sweep wick
                    const risk_distance = entry_price - sl;
                    tp = entry_price + (risk_distance * 4.0); // 1:4 R/R Expansion
                    
                    const risk_amount = balance * 0.025;
                    trade_volume = risk_amount / risk_distance;
                    has_traded_today = true;
                }
            }
        }

        // Manage Trade
        if (in_trade) {
            if (trade_dir == 1) { // Buy
                if (c.low <= sl) {
                    balance -= (entry_price - sl) * trade_volume;
                    in_trade = false;
                    total_trades += 1;
                } else if (c.high >= tp) {
                    balance += (tp - entry_price) * trade_volume;
                    in_trade = false;
                    total_trades += 1;
                    wins += 1;
                }
            } else if (trade_dir == -1) { // Sell
                if (c.high >= sl) {
                    balance -= (sl - entry_price) * trade_volume;
                    in_trade = false;
                    total_trades += 1;
                } else if (c.low <= tp) {
                    balance += (entry_price - tp) * trade_volume;
                    in_trade = false;
                    total_trades += 1;
                    wins += 1;
                }
            }
        }
        
        if (balance > max_balance) max_balance = balance;
    }

    std.debug.print("=========================================================\n", .{});
    std.debug.print("          [KESSLER V1.1] ZIG BARE-METAL BACKTEST         \n", .{});
    std.debug.print("               EDGE: ASIAN LIQUIDITY SWEEP (AMD)         \n", .{});
    std.debug.print("=========================================================\n", .{});
    std.debug.print("Starting Equity: $10,750.00 (₹9 Lakh INR)\n", .{});
    std.debug.print("Ending Equity:   ${d:.2}\n", .{balance});
    std.debug.print("Total Profit:    ${d:.2}\n", .{balance - 10750.0});
    std.debug.print("Max Equity:      ${d:.2}\n", .{max_balance});
    std.debug.print("Total Sweeps:    {}\n", .{total_trades});
    if (total_trades > 0) {
        const win_rate: f64 = @as(f64, @floatFromInt(wins)) / @as(f64, @floatFromInt(total_trades)) * 100.0;
        std.debug.print("Sniper Win Rate: {d:.1}%\n", .{win_rate});
    }
    std.debug.print("=========================================================\n", .{});
}

File: research_and_backtests/amd_backtest.py
import pandas as pd
import numpy as np

def run_backtest():
    try:
        df = pd.read_csv("xauusd_clean.csv", names=["time", "open", "high", "low", "close"])
    except Exception as e:
        print(f"Error reading CSV: {e}")
        return

    print("=========================================================")
    print("        [KESSLER V1.1] AMD EDGE BACKTEST (60 DAYS)       ")
    print("=========================================================")
    print(f"[*] Python Fallback Engine: Loaded {len(df)} M5 Candles.")
    
    balance = 10750.0  # $10,750 USD = 9 Lakh INR
    max_balance = balance
    total_trades = 0
    wins = 0

    in_trade = False
    trade_dir = 0
    entry_price = 0
    sl = 0
    tp = 0
    trade_volume = 0

    asian_high = 0.0
    asian_low = 99999.0
    has_traded_today = False
    current_day = -1

    # Convert time to datetime
    df['time'] = pd.to_datetime(df['time'], unit='s')

    swept_high = False
    swept_low = False

    for index, row in df.iterrows():
        dt = row['time']
        day = dt.day
        hour = dt.hour

        if day != current_day:
            current_day = day
            asian_high = 0.0
            asian_low = 99999.0
            has_traded_today = False
            swept_high = False
            swept_low = False
            
            # Close trade overnight
            if in_trade:
                if trade_dir == 1:
                    balance += (row['open'] - entry_price) * trade_volume
                else:
                    balance += (entry_price - row['open']) * trade_volume
                in_trade = False

        # Asian Session (01:00 to 09:00 Broker Time)
        if 1 <= hour < 9:
            if row['high'] > asian_high: asian_high = row['high']
            if row['low'] < asian_low: asian_low = row['low']

        # London/NY Session (09:00 to 22:00)
        if 9 <= hour < 22 and not in_trade:
            range_size = asian_high - asian_low
            
            if 3.0 < range_size < 45.0:
                # Track sweeps
                if row['high'] > asian_high: swept_high = True
                if row['low'] < asian_low: swept_low = True

                # Sell Signal: Fade the Breakout of Asian High (Mean Reversion)
                if row['close'] > (asian_high + 0.5):
                    in_trade = True
                    trade_dir = -1
                    entry_price = row['close']
                    sl = entry_price + 6.0  # 60 pip stop
                    risk_dist = sl - entry_price
                    tp = entry_price - 12.0 # 1:2 R/R
                    
                    risk_amount = balance * 0.03
                    trade_volume = risk_amount / risk_dist if risk_dist > 0 else 0
                    has_traded_today = True

                # Buy Signal: Fade the Breakout of Asian Low (Mean Reversion)
                elif row['close'] < (asian_low - 0.5):
                    in_trade = True
                    trade_dir = 1
                    entry_price = row['close']
                    sl = entry_price - 6.0 # 60 pip stop
                    risk_dist = entry_price - sl
                    tp = entry_price + 12.0 # 1:2 R/R
                    
                    risk_amount = balance * 0.03
                    trade_volume = risk_amount / risk_dist if risk_dist > 0 else 0
                    has_traded_today = True

        if in_trade:
            if trade_dir == 1:
                if row['low'] <= sl:
                    balance -= (entry_price - sl) * trade_volume
                    in_trade = False
                    total_trades += 1
                elif row['high'] >= tp:
                    balance += (tp - entry_price) * trade_volume
                    in_trade = False
                    total_trades += 1
                    wins += 1
            else:
                if row['high'] >= sl:
                    balance -= (sl - entry_price) * trade_volume
                    in_trade = False
                    total_trades += 1
                elif row['low'] <= tp:
                    balance += (entry_price - tp) * trade_volume
                    in_trade = False
                    total_trades += 1
                    wins += 1
                    
        if balance > max_balance:
            max_balance = balance

    print("=========================================================")
    print(f"Starting Equity: $10,750.00 (₹9 Lakh INR)")
    print(f"Ending Equity:   ${balance:.2f}")
    print(f"Total Profit:    ${balance - 10750.0:.2f}")
    print(f"Max Equity:      ${max_balance:.2f}")
    print(f"Total Sweeps:    {total_trades}")
    if total_trades > 0:
        print(f"Sniper Win Rate: {(wins/total_trades)*100:.1f}%")
    print("=========================================================")

if __name__ == "__main__":
    run_backtest()
